File: challenge_C/level2/삼각 달팽이.c
#include <stdio.h>
#include <stdlib.h>

int limit, idx, max;
int num = 1, cnt = 1;
int *ans;

void plus_idx()
{
    for(; cnt < limit; cnt++)
    {
        ans[idx] = num++;
        idx += cnt;
    }
    
    if(cnt == limit)
    {
        for(; idx < max && !ans[idx]; idx++)
            ans[idx] = num++;
        idx--;
    }
}

void minus_idx()
{
    for(; !ans[idx - cnt]; cnt--)
    {
        idx -= cnt;
        ans[idx] = num++;
    }
}

int* solution(int n)
{
    limit = n;
    max = n * (n + 1) / 2;
    ans = (int*)calloc(max, sizeof(int));
    
    while(num <= max)
    {
        plus_idx();
        minus_idx();
        idx += cnt++;
        limit--;
    }
    
    return ans;
}

File: challenge_Java/level2/삼각 달팽이.java
class Solution {
    public int[][] map;
    public int num = 1, cnt, row = 0, col = 0;
    
    public int[] solution(int n) {
        cnt = n;
        map = new int[n][n];
        
        int len = (n + 1) * (n / 2);
        len += n % 2 == 0 ? 0 : n / 2 + 1;
        
        while (num <= len) {
            increaseRow();
            increaseCol();
            decreaseRowCol();
        }
        
        int idx = 0;
        int[] answer = new int[len];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (map[i][j] == 0)
                    break;
                answer[idx++] = map[i][j];
            }
        }
        return answer;
    }
    
    public void increaseRow() {
        for (int i = 0; i < cnt; i++) {
            map[row++][col] = num++;
        }
        row--;
        col++;
        cnt--;
    }
    
    public void increaseCol() {
        for (int i = 0; i < cnt; i++) {
            map[row][col++] = num++;
        }
        col -= 2;
        row--;
        cnt--;
    }
    
    public void decreaseRowCol() {
        for (int i = 0; i < cnt; i++) {
            map[row--][col--] = num++;
        }
        row += 2;
        col++;
        cnt--;
    }
}
